cpp/src/decisiontree/batched-levelalgo/kernels.cuh

```diff
@@ -362,50 +367,126 @@ __global__ void computeSplitClassificationKernel(
   for (auto i = range_start + tid; i < end; i += stride) {
+    // each thread works over a data point and strides to the next
     auto row = input.rowids[i];
     auto d = input.data[row + coloffset];
     auto label = input.labels[row];
+    // breakflag to check if data is greater than largest bin
+    bool breakflag = false;
     for (IdxT b = 0; b < nbins; ++b) {
-      auto isRight = d > sbins[b];  // no divergence
-      auto offset = b * 2 * nclasses + isRight * nclasses + label;
-      atomicAdd(shist + offset, 1);  // class hist
+      if(d <= sbins[b]) { // shist (0 -> nbins*nclasses - 1)
+        auto offset = b * nclasses + label;
+        atomicAdd(pdf_shist + offset, 1);
+        breakflag = true;
+        break;
+      }
     }
+    // case when d is larger than all bins
+    if(!breakflag) atomicAdd(pdf_shist + nbins*nclasses + label, 1);
   }
+
+  // synchronizeing above changes across block
   __syncthreads();
+  __threadfence();
+
   // update the corresponding global location
-  auto histOffset = ((nid * gridDim.y) + blockIdx.y) * len;
-  for (IdxT i = threadIdx.x; i < len; i += blockDim.x) {
-    atomicAdd(hist + histOffset + i, shist[i]);
+  auto histOffset = ((nid * gridDim.y) + blockIdx.y) * pdf_shist_len;
+  for (IdxT i = threadIdx.x; i < pdf_shist_len; i += blockDim.x) {
+    atomicAdd(hist + histOffset + i, pdf_shist[i]);
   }
+
   __threadfence();  // for commit guarantee
   __syncthreads();
+
   // last threadblock will go ahead and compute the best split
   bool last = true;
   if (gridDim.x > 1) {
     last = MLCommon::signalDone(done_count + nid * gridDim.y + blockIdx.y,
                                 gridDim.x, blockIdx.x == 0, sDone);
   }
+  // if not the last threadblock, exit
   if (!last) return;
-  for (IdxT i = threadIdx.x; i < len; i += blockDim.x)
-    shist[i] = hist[histOffset + i];
+
+  // store the complete global histogram in shared memory of last block
+  for (IdxT i = threadIdx.x; i < pdf_shist_len; i += blockDim.x)
+    pdf_shist[i] = hist[histOffset + i];
+
+  __syncthreads();
+
+  // Blockscan instance preparation
+  typedef cub::BlockScan<int, TPB> BlockScan;
+  __shared__ typename BlockScan::TempStorage temp_storage;
+
+  /**
+   * Scanning code:
+   * span: block-wide
+   * Function: convert the PDF calculated in the previous steps to CDF
+   * This CDF is done over 2 passes
+   * * one from left to right to populate counts of left splits
+   *   for each split-point.
+   * * second from right to left to populate the right splits
+   *   for each split-point
+   */
+  for(IdxT tix = threadIdx.x; tix < max(TPB, nbins); tix += blockDim.x) {
+    for(IdxT c = 0; c < nclasses; ++c){
+      // for each class, do inclusive block scan
+      int pdf_per_bin_per_class;
+      int cdf_per_bin_per_class;
+      // left to right scan operation for scanning lesser-than-or-equal-to-bin counts
+      // offset for left to right scan of pdf_shist
+      IdxT class_segment_offset = nclasses*tix;
```

Comment:
this could now lead to smem bank conflicts (as a function of `nclasses` ofcourse!). So, I was wondering if we should change the `pdf_shist` and `cdf_shist` to instead be arranged as `nclasses x nbins`. Currently these are organized as `nbins x nclasses`. What do you think, @vinaydes ?

Reply:
this also means we might have to modify the metrics computation!

Reply:
@venkywonka Is it possible to get numbers on shared memory bank conflicts before and after this change?

Reply:
Yes I understand the problem 🤔 So on an average there is a 2x increase in smem read conflicts before to after change. Write conflicts are practically non-existent in old code, but are comparable to read bank conflicts after change...

Reply:
@teju85 , have arranged access patterns according to `nclasses x nbins` in new changes, they do seem to reduce write and read conflicts

Reply:
Thanks.